src/coroutine/memoset/mod.rs

```diff
@@ -271,54 +271,63 @@ impl<F: LurkField, Q: Query<F>> Scope<Q, LogMemo<F>> {
     fn build_transcript(&self, s: &Store<F>) -> (Transcript<F>, Vec<Ptr>) {
         let mut transcript = Transcript::new(s);
 
-        let internal_insertions_kv = self.internal_insertions.iter().map(|key| {
-            let value = self.queries.get(key).expect("value missing for key");
-            Transcript::make_kv(s, *key, *value)
-        });
+        // k -> [kv]
+        let mut insertions: HashMap<Ptr, IndexSet<Ptr>> = HashMap::new();
+        let mut unique_keys: Vec<Ptr> = Default::default();
 
-        let mut insertions =
-            Vec::with_capacity(self.toplevel_insertions.len() + self.internal_insertions.len());
-        insertions.extend(&self.toplevel_insertions);
-        insertions.extend(internal_insertions_kv);
+        let mut insert = |kv: Ptr| {
+            let key = s.car_cdr(&kv).unwrap().0;
 
-        // Sort insertions by query type (index) for processing. This is because the transcript will be constructed
-        // sequentially by the circuits, and we potentially batch queries of the same type in a single coprocessor
-        // circuit.
-        insertions.sort_by_key(|kv| {
-            let (key, _) = s.car_cdr(kv).unwrap();
+            if let Some(kvs) = insertions.get_mut(&key) {
+                kvs.insert(kv);
+            } else {
+                unique_keys.push(key);
+                let mut x = IndexSet::new();
+                x.insert(kv);
 
-            Q::from_ptr(s, &key).expect("invalid query").index()
+                insertions.insert(key, x);
+            }
+        };
+
+        let internal_insertions_kv = self.internal_insertions.iter().map(|key| {
+            let value = self.queries.get(key).expect("value missing for key");
+            Transcript::make_kv(s, *key, *value)
         });
 
+        for kv in &self.toplevel_insertions {
+            insert(*kv);
+        }
+        for kv in internal_insertions_kv {
+            insert(kv);
+        }
         for kv in self.toplevel_insertions.iter() {
             transcript.add(s, *kv);
         }
 
+        // Sort insertion keys by query type (index) for processing. This is because the transcript will be constructed
+        // sequentially by the circuits, and we potentially batch queries of the same type in a single coprocessor
+        // circuit.
+        unique_keys.sort_by_key(|k| Q::from_ptr(s, k).expect("invalid query").index());
```

Comment:
It's worth noting this PR doesn't (yet) eliminate all sorting — just improves the situation. I'm working on separating by query index from the beginning so we can avoid this also. That doesn't necessarily need to land in this PR, but if I complete before this merges, I'll push here.

Reply:
Fixed here: https://github.com/lurk-lab/lurk-rs/pull/1057/commits/7592f1cb0317038646013d14e942e1813d841700